File: sdk/nodejs/src/exceptions.ts
/**
 * Exception classes for AgentCache SDK
 */

/**
 * Base exception for all AgentCache errors
 */
export class AgentCacheError extends Error {
  public statusCode?: number;
  public response?: any;

  constructor(message: string, statusCode?: number, response?: any) {
    super(message);
    this.name = 'AgentCacheError';
    this.statusCode = statusCode;
    this.response = response;
    Object.setPrototypeOf(this, AgentCacheError.prototype);
  }
}

/**
 * Raised when API key is invalid or missing
 */
export class AuthenticationError extends AgentCacheError {
  constructor(message: string = 'Invalid or missing API key') {
    super(message, 401);
    this.name = 'AuthenticationError';
    Object.setPrototypeOf(this, AuthenticationError.prototype);
  }
}

/**
 * Raised when API rate limit is exceeded
 */
export class RateLimitError extends AgentCacheError {
  public retryAfter?: number;

  constructor(message: string = 'Rate limit exceeded', retryAfter?: number) {
    super(message, 429);
    this.name = 'RateLimitError';
    this.retryAfter = retryAfter;
    Object.setPrototypeOf(this, RateLimitError.prototype);
  }
}

/**
 * Raised when request validation fails
 */
export class ValidationError extends AgentCacheError {
  constructor(message: string = 'Invalid request parameters') {
    super(message, 400);
    this.name = 'ValidationError';
    Object.setPrototypeOf(this, ValidationError.prototype);
  }
}

/**
 * Raised when requested resource is not found
 */
export class NotFoundError extends AgentCacheError {
  constructor(message: string = 'Resource not found') {
    super(message, 404);
    this.name = 'NotFoundError';
    Object.setPrototypeOf(this, NotFoundError.prototype);
  }
}

/**
 * Raised when server encounters an error
 */
export class ServerError extends AgentCacheError {
  constructor(message: string = 'Internal server error') {
    super(message, 500);
    this.name = 'ServerError';
    Object.setPrototypeOf(this, ServerError.prototype);
  }
}

/**
 * Raised when request times out
 */
export class TimeoutError extends AgentCacheError {
  constructor(message: string = 'Request timeout') {
    super(message, 408);
    this.name = 'TimeoutError';
    Object.setPrototypeOf(this, TimeoutError.prototype);
  }
}

/**
 * Raised when network connection fails
 */
export class NetworkError extends AgentCacheError {
  constructor(message: string = 'Network connection failed') {
    super(message);
    this.name = 'NetworkError';
    Object.setPrototypeOf(this, NetworkError.prototype);
  }
}

File: sdk/python/agentcache/exceptions.py
"""Exception classes for AgentCache SDK."""


class AgentCacheError(Exception):
    """Base exception for all AgentCache errors."""
    
    def __init__(self, message: str, status_code: int = None, response: dict = None):
        self.message = message
        self.status_code = status_code
        self.response = response
        super().__init__(self.message)


class AuthenticationError(AgentCacheError):
    """Raised when API key is invalid or missing."""
    
    def __init__(self, message: str = "Invalid or missing API key"):
        super().__init__(message, status_code=401)


class RateLimitError(AgentCacheError):
    """Raised when API rate limit is exceeded."""
    
    def __init__(self, message: str = "Rate limit exceeded", retry_after: int = None):
        super().__init__(message, status_code=429)
        self.retry_after = retry_after


class ValidationError(AgentCacheError):
    """Raised when request validation fails."""
    
    def __init__(self, message: str = "Invalid request parameters"):
        super().__init__(message, status_code=400)


class NotFoundError(AgentCacheError):
    """Raised when requested resource is not found."""
    
    def __init__(self, message: str = "Resource not found"):
        super().__init__(message, status_code=404)


class ServerError(AgentCacheError):
    """Raised when server encounters an error."""
    
    def __init__(self, message: str = "Internal server error"):
        super().__init__(message, status_code=500)


class TimeoutError(AgentCacheError):
    """Raised when request times out."""
    
    def __init__(self, message: str = "Request timeout"):
        super().__init__(message, status_code=408)


class NetworkError(AgentCacheError):
    """Raised when network connection fails."""
    
    def __init__(self, message: str = "Network connection failed"):
        super().__init__(message)
